src/gc/gcinterface.ee.h

```diff
@@ -0,0 +1,150 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#ifndef _GCINTERFACE_EE_H_
+#define _GCINTERFACE_EE_H_
+
+struct ScanContext;
+struct gc_alloc_context;
+class CrawlFrame;
+
+// Callback passed to GcScanRoots.
+typedef void promote_func(PTR_PTR_Object, ScanContext*, uint32_t);
+
+// Callback passed to GcEnumAllocContexts.
+typedef void enum_alloc_context_func(gc_alloc_context*, void*);
+
+// Callback passed to CreateBackgroundThread.
+typedef uint32_t (__stdcall *GCBackgroundThreadFunction)(void* param);
+
+// Struct often used as a parameter to callbacks.
+typedef struct
+{
+    promote_func*  f;
+    ScanContext*   sc;
+    CrawlFrame *   cf;
+} GCCONTEXT;
+
+// SUSPEND_REASON is the reason why the GC wishes to suspend the EE,
+// used as an argument to IGCToCLR::SuspendEE.
+typedef enum
+{
+    SUSPEND_FOR_GC = 1,
+    SUSPEND_FOR_GC_PREP = 6
+} SUSPEND_REASON;
+
+// The rationale behind these defines it that, if we are intending
+// to link the GC statically to the runtime, there's no need to dispatch
+// this interface virtually and we can call directly into the interface
+// implementation. If our GC will be dynamically linked, though, this
+// interface must be virtual.
+#ifdef FEATURE_STANDALONE_GC
+# define EE_CALL_DECL_BEGIN virtual
+# define EE_CALL_DECL_END = 0
+# define EE_INTERFACE_NAME IGCToCLR
+#else
+# define EE_CALL_DECL_BEGIN static
+# define EE_CALL_DECL_END
+# define EE_INTERFACE_NAME GCToEEInterface
+
+// Even if FEATURE_STANDALONE_GC is not defined,
+// this type still exists as an opaque class.
+class IGCToCLR;
+
+#endif // FEATURE_STANDALONE_GC
+
+// This interface provides the interface that the GC will use to speak to the rest
+// of the execution engine. Everything that the GC does that requires the EE
+// to be informed or that requires EE action must go through this interface.
+//
+// When FEATURE_STANDALONE_GC is defined, this class is named IGCToCLR and is
+// an abstract class. The EE will provide a class that fulfills this interface,
+// and the GC will dispatch virtually on it to call into the EE. When FEATURE_STANDALONE_GC
+// is not defined, this class is named GCToEEInterface and the GC will dispatch statically on it.
+class EE_INTERFACE_NAME {
```

Comment:
Just to be sure I'm understanding this right:
* The `GCToEEInterface` in `gcenv.ee.h` (as it is today) will be defined when `FEATURE_STANDALONE_GC` is not defined,
* The `GCToEEInterface` in `gctoeeinterface.h` (the one forwarding to `IGCToCLR`) will be defined when `FEATURE_STANDALONE_GC` _is_ defined, and
* Either way, `GCToEEInterface` will be implemented in `gcenv.ee.cpp`, either deriving from `IGCToCLR` if `FEATURE_STANDALONE_GC` is defined or as a static class when it is not defined.

Also by leaving the `GCToEEInterface` as it is today, we decouple the declaration of `IGCToCLR` from `GCToEEInterface`. Does that sound correct?